pageserver/src/tenant/timeline.rs

```diff
@@ -3643,194 +3578,251 @@ impl Timeline {
 
-            // Fsync all the layer files and directory using multiple threads to
-            // minimize latency.
-            par_fsync::par_fsync(&layer_paths).context("fsync all new layers")?;
+                // Fsync all the layer files and directory using multiple threads to
+                // minimize latency.
+                par_fsync::par_fsync(&layer_paths).context("fsync all new layers")?;
 
-            par_fsync::par_fsync(&[self.conf.timeline_path(&self.timeline_id, &self.tenant_id)])
+                par_fsync::par_fsync(&[myself
+                    .conf
+                    .timeline_path(&myself.timeline_id, &myself.tenant_id)])
                 .context("fsync of timeline dir")?;
 
-            layer_paths.pop().unwrap();
-        }
+                layer_paths.pop().unwrap();
+            }
 
-        stats.write_layer_files_micros = stats.sort_holes_micros.till_now();
-        stats.new_deltas_count = Some(new_layers.len());
-        stats.new_deltas_size = Some(new_layers.iter().map(|l| l.desc.file_size).sum());
+            stats.write_layer_files_micros = stats
+                .read_lock_drop_until_spawn_blocking_code_start_micros
+                .till_now();
+            stats.new_deltas_count = Some(new_layers.len());
+            stats.new_deltas_size = Some(new_layers.iter().map(|l| l.desc.file_size).sum());
 
-        drop(all_keys_iter); // So that deltas_to_compact is no longer borrowed
+            drop(all_keys_iter); // So that deltas_to_compact is no longer borrowed
 
-        match TryInto::<CompactLevel0Phase1Stats>::try_into(stats)
-            .and_then(|stats| serde_json::to_string(&stats).context("serde_json::to_string"))
-        {
-            Ok(stats_json) => {
-                info!(
-                    stats_json = stats_json.as_str(),
-                    "compact_level0_phase1 stats available"
-                )
-            }
-            Err(e) => {
-                warn!("compact_level0_phase1 stats failed to serialize: {:#}", e);
+            match TryInto::<CompactLevel0Phase1Stats>::try_into(stats)
+                .and_then(|stats| serde_json::to_string(&stats).context("serde_json::to_string"))
+            {
+                Ok(stats_json) => {
+                    info!(
+                        stats_json = stats_json.as_str(),
+                        "compact_level0_phase1 stats available"
+                    )
+                }
+                Err(e) => {
+                    warn!("compact_level0_phase1 stats failed to serialize: {:#}", e);
+                }
             }
-        }
 
-        Ok(CompactLevel0Phase1Result {
-            new_layers,
-            deltas_to_compact,
+            Ok(CompactLevel0Phase1Result {
+                new_layers,
+                deltas_to_compact,
+            })
         })
+        .await
+        .context("spawn_blocking")
+        .map_err(CompactionError::Other)??;
```

Comment:
It took me some time to understand the original `DownloadError` can be passed as-is here given this is two layers of error wrapping

Reply:
... and `CompactionError::Other` is only applied on the spawn_blocking join error

Reply:
Yeah, I don't like it either. I'm currently juggling the code around a bit more, this will go away

Reply:
https://github.com/neondatabase/neon/pull/4537/commits/9e1de941bd902a6bd5d098e70a959a58c94a923d